sycl/doc/extensions/proposed/sycl_ext_oneapi_enqueue_functions.asciidoc

```diff
@@ -0,0 +1,458 @@
+not required to describe application behavior.
+
+All functions proposed in this extension accept as their first argument an
+object that represents where a command should be submitted. For now, we refer
+to this as an "Executor". This argument can currently be either a
+`sycl::handler` or a `sycl::queue`, allowing the new functions to be used
+either at command-group scope or as a replacement for existing queue shortcuts.
+It is expected that a future version of this extension will adjust this
+overload set to include functions compatible with future C++ concepts (such as
+senders and receivers).
+
+
+=== Usage example
+
+The example below demonstrates that the syntax proposed here requires only
+minor changes to existing applications, while retaining their structure.
+
+
+==== SYCL 2020
+
+[source,c++]
+----
+q.submit([&](sycl::handler& h) {
+  sycl::accessor result { buf, h, sycl::write_only, sycl::no_init };
+  h.parallel_for(1024, [=](sycl::id<1> idx) {
+    result[idx] = idx;
+  });
+});
+
+std::vector<sycl::event> depEvents = /* some dependencies */;
+sycl::event e = q.parallel_for(sycl::nd_range<1>{1024, 16}, depEvents,
+  [=](sycl::nd_item<1> it) {
+  result[it.get_global_id()] += 1;
+});
+e.wait();
+----
+
+
+==== Proposed syntax
+
+[source,c++]
+----
+using syclex = sycl::ext::oneapi::experimental;
+
+syclex::submit(q, [&](sycl::handler& h) {
+  sycl::accessor result { buf, h, sycl::write_only, sycl::no_init };
+  syclex::parallel_for(h, 1024, [=](sycl::id<1> idx) {
+    result[idx] = idx;
+  });
+});
+
+std::vector<sycl::event> depEvents = /* some dependencies */;
+sycl::event e;
+syclex::nd_launch(q, sycl::nd_range<1>{1024, 16}, [=](sycl::nd_item<1> it) {
+  result[it.get_global_id()] += 1;
+}, depEvents, &e);
+e.wait();
+----
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_ENQUEUE_FUNCTIONS` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Command group submission
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename ExecutorType, typename CommandGroupFunc>
+void submit(ExecutorType ex, CommandGroupFunc cgf, const std::vector<event>& depEvents = {}, event* outEvent = nullptr);
+
+}
+----
+
+[source,c++]
+----
+template <typename ExecutorType, typename CommandGroupFunc>
+void submit(ExecutorType ex, CommandGroupFunc cgf, const std::vector<event>& depEvents = {}, event* outEvent = nullptr);
```

Comment:
Do we want perfect forwarding instead like this?

```suggestion
void submit(ExecutorType ex, CommandGroupFunc &&cgf, const std::vector<event>& depEvents = {}, event* outEvent = nullptr);
```

I think @rolandschulz argued previously that it was a mistake for `queue::submit` to take the callable by value.

Reply:
Added in 756483fb36d52fa29a0c96f7325aea000a858e0a.